lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -262,33 +263,38 @@ replaceLoad(LoadInst *li, SILValue newValue, AllocStackInst *asi,
     // to guaranteed!
     if (proj.getKind() == ProjectionKind::Struct ||
         proj.getKind() == ProjectionKind::Tuple) {
-      if (auto opVal = scope.borrowValue(li, newValue)) {
+      if (auto opVal = scope.borrowValue(inst, newValue)) {
         assert(*opVal != newValue &&
                "Valid value should be different from input value");
         newValue = *opVal;
       }
     }
     newValue =
-        proj.createObjectProjection(builder, li->getLoc(), newValue).get();
+        proj.createObjectProjection(builder, inst->getLoc(), newValue).get();
   }
 
-  op = li->getOperand();
+  op = inst->getOperand(0);
 
-  // Replace users of the loaded value with `val`
-  // If we have a load [copy], replace the users with copy_value of `val`
-  if (li->getOwnershipQualifier() == LoadOwnershipQualifier::Copy) {
-    li->replaceAllUsesWith(builder.createCopyValue(li->getLoc(), newValue));
+  if (auto *li = dyn_cast<LoadInst>(inst)) {
+    // Replace users of the loaded value with `newVal`
+    // If we have a load [copy], replace the users with copy_value of `newVal`
+    if (li->getOwnershipQualifier() == LoadOwnershipQualifier::Copy) {
+      li->replaceAllUsesWith(builder.createCopyValue(li->getLoc(), newValue));
+    } else {
+      li->replaceAllUsesWith(newValue);
+    }
   } else {
-    assert(!asi->getFunction()->hasOwnership() ||
-           newValue.getOwnershipKind() != OwnershipKind::Guaranteed);
-    li->replaceAllUsesWith(newValue);
+    auto *lbi = cast<LoadBorrowInst>(inst);
+    auto *borrow = SILBuilderWithScope(lbi, ctx).createBeginBorrow(
```

Comment:
Probably not an issue:

For lexical alloc_stacks, stores get transformed to 
```
%lifetime = begin_borrow [lexical] %original
%copy = copy_value %lifetime
```

Would it be better for all uses of `lbi` to be replaced with `%lifetime` rather than this `begin_borrow [lexical] %copy`?  Or alternatively for this inner borrow to be non-lexical, considering that the outer lexical borrow scope (which replaces the range of instructions `store...destroy_addr` where this load_borrow and its uses appear) already specifies the lifetime?

Reply:
yep. A lexical scope is supposed to start at the point at which a variable was assigned and end at the end of scope. Since the load_borrow was only there to keep the memory location alive, we shouldn't need a borrow anymore. Is it easy to just reuse the incoming `begin_borrow` and delete the existing `end_borrow`s?

Reply:
I will try that out, I am not sure if the current algorithm will take care of adding lexical lifetime ends in the presence of reborrows.

Reply:
Updated PR to replace by the lexical borrow. Disabling the optimization when there are reborrows for now. Since the mem2reg algorithm just adds the new lifetime and lexical borrows at phi blocks, the current algorithm will not work as is when there are reborrows.

Reply:
Just skimmed the diff again. LGTM